renderdoc/driver/vulkan/wrappers/vk_sync_funcs.cpp

```diff
@@ -1505,6 +1505,124 @@ void WrappedVulkan::vkCmdWaitEvents2(VkCommandBuffer commandBuffer, uint32_t eve
+  if(transitionCount > 0)
+    ser.Important();
+
+  Serialise_DebugMessages(ser);
+
+  SERIALISE_CHECK_READ_ERRORS();
+
+  rdcarray<VkHostImageLayoutTransitionInfo> imgBarriers;
+
+  if(IsReplayingAndReading())
+  {
+    for(uint32_t i = 0; i < transitionCount; i++)
+    {
+      if(pTransitions[i].image != VK_NULL_HANDLE)
+      {
+        VkImage image = pTransitions[i].image;
+
+        imgBarriers.push_back(pTransitions[i]);
+        imgBarriers.back().image = Unwrap(image);
+
+        SanitiseOldImageLayout(imgBarriers.back().oldLayout);
+        SanitiseReplayImageLayout(imgBarriers.back().newLayout);
+
+        if(IsActiveReplaying(m_State) &&
+           m_ReplayOptions.optimisation != ReplayOptimisationLevel::Fastest)
+        {
+          if(pTransitions[i].oldLayout == VK_IMAGE_LAYOUT_UNDEFINED)
+          {
+            VkImageLayout newLayout = pTransitions[i].newLayout;
+            SanitiseNewImageLayout(newLayout);
+
+            GetDebugManager()->FillWithDiscardPatternOnHost(
+                device, DiscardType::UndefinedTransition, image, newLayout,
+                pTransitions[i].subresourceRange, {{0, 0}, {65536, 65536}});
+          }
+        }
+      }
+    }
+
+    ObjDisp(device)->TransitionImageLayoutEXT(Unwrap(device), (uint32_t)imgBarriers.size(),
+                                              imgBarriers.data());
+  }
+
+  return true;
+}
+
+VkResult WrappedVulkan::vkTransitionImageLayoutEXT(VkDevice device, uint32_t transitionCount,
+                                                   const VkHostImageLayoutTransitionInfo *pTransitions)
+{
+  SCOPED_DBG_SINK();
+
+  VkResult ret;
+  {
+    byte *memory = GetTempMemory(sizeof(VkHostImageLayoutTransitionInfo) * transitionCount);
+    VkHostImageLayoutTransitionInfo *im = (VkHostImageLayoutTransitionInfo *)memory;
+
+    for(uint32_t i = 0; i < transitionCount; i++)
+    {
+      im[i] = pTransitions[i];
+      im[i].image = Unwrap(im[i].image);
+    }
+
+    SERIALISE_TIME_CALL(
+        ret = ObjDisp(device)->TransitionImageLayoutEXT(Unwrap(device), transitionCount, im));
+  }
+
+  if(IsActiveCapturing(m_State))
+  {
+    CACHE_THREAD_SERIALISER();
+
+    SCOPED_SERIALISE_CHUNK(VulkanChunk::vkTransitionImageLayout);
+    Serialise_vkTransitionImageLayoutEXT(ser, device, transitionCount, pTransitions);
+
+    m_FrameCaptureRecord->AddChunk(scope.Get());
+
+    for(uint32_t i = 0; i < transitionCount; ++i)
+    {
+      const VkHostImageLayoutTransitionInfo &transition = pTransitions[i];
+
+      // Since this transition is done on the host, record a barrier for this image and immediately
+      // apply it.
+      rdcflatmap<ResourceId, ImageState> imageStates;
+      imageStates.insert(
+          {GetResID(transition.image),
+           ImageState(
+               transition.image,
+               GetResourceManager()->GetResourceRecord(GetResID(transition.image))->resInfo->imageInfo,
+               eFrameRef_None)});
```

Comment:
Can you go into detail about why this initial insert is here? I'm guessing you added this because you ran into a problem recording the barrier normally but I don't see the command buffer recording of barriers doing anything similar - this _should_ be handled inside `RecordBarriers` when it encounters a barrier for a new image and it goes and fetches the initial image state. It feels like it should be fixed a different way but I'm not sure what the problem is that prompted it.